DonutMeter Component: Set up a reusable donut chart component

annacmc opened the pull request:
This PR adds a DonutMeter  component to Jetpack reusable components. 

Fixes #25917

#### Changes proposed in this Pull Request:

* Adds `DonutMeter` Component 
* Adds the component to storybook
* Ensures component takes props for adjusting the size, chart thickness, chart segment color, current value, max/total value
* Ensures component has an accessible/screen reader table alternative
* Ensures chart segment begins at the bottom (180 degrees) of the donut and increments in a clockwise direction. 

![image](https://user-images.githubusercontent.com/30754158/187625076-0d2277b7-888b-45d3-9a1e-0e8386dbe351.png)


#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?


#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:

* Run `cd projects/js-packages/storybook && npm run storybook:dev` and confirm that you see DonutMeter component in Storybook.
* Under JS Packages → Components → DonutMeter, choose the story Default and try changing some of the variables under Controls. You should be able to adjust the chart size, donut line width, segment colour, the segment count, and the total count.

You can also run the unit tests for js-packages/components by running `jetpack test`.

annacmc committed:
[not verified] set up basic donut component

annacmc committed:
[not verified] add donut chart to story book

annacmc committed:
[not verified] add basic donut styles

annacmc committed:
[not verified] pass value as percentage

annacmc committed:
[not verified] remove unnecessary styling

annacmc committed:
[not verified] pass background color to chart segment

annacmc committed:
[not verified] add chart thickness prop

annacmc committed:
[not verified] add default args

annacmc committed:
[not verified] add accessible screenreader table

annacmc committed:
[not verified] tidy up styling

annacmc committed:
[not verified] add chart size (width) control

annacmc committed:
[not verified] remove label prop

annacmc committed:
[not verified] changelogger

annacmc committed:
[not verified] empty stylesheet for now

annacmc committed:
[not verified] tools/fixup-project-versions.sh

annacmc committed:
tools/fixup-project-versions.sh

annacmc committed:
one more try at tools/fixup-project-versions.sh

annacmc committed:
add test dir & file

annacmc committed:
fix up naming

annacmc committed:
add basic svg render test

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+import { __ } from '@wordpress/i18n';
+import React from 'react';
+import './style.scss';
+export type RecordMeterDonutProps = {
```
Could we name it something more general, because the donut meter will also be used for as search request meter. And in the future, possibly sharing meter etc 😄 

This also applies to some of the naming of CSS classes etc.

annacmc replied:
sure - that's totally doable. I just named it after its sibling RecordMeterBar component. What about DonutMeter or DonutChart?

annacmc replied:
Okay, gone with DonutMeter for now, seems a bit more general use to me :)

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+					className="record-meter-donut-segment"
+					cx="20"
+					cy="20"
+					r="15.91549430918954"
```
I'm curious - where does this number come from?

annacmc replied:
the `r` number? this is the radius of the circle. The logic comes from [this blog post](https://heyoka.medium.com/scratch-made-svg-donut-pie-charts-in-html5-2c587e935d72) which says:

>radius based on the circumference is: r = C/(2π), or r = 100/(2π). The result is 15.91549430918952. This is a crazy number with lots of decimals, I know, but it will make all of our lives easier later.

annacmc replied:
I've added some comments into the code to help :)

kangzj commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+					stroke={ backgroundColor }
+					strokeWidth={ thickness }
+					strokeDasharray={ `${ count() } ${ 100 - count() }` }
+					strokeDashoffset="-25" // this ensures the segment begins at the bottom of the donut instead of the top
```
👍

kangzj reviewed (commented):
It looks nice - sorry only left a couple of comments and have to hop off - will continue tomorrow.

annacmc committed:
add comments

annacmc committed:
rename RecordMeterDonut to DonutMeter

annacmc commented:
> It looks nice - sorry only left a couple of comments and have to hop off - will continue tomorrow.

thanks @kangzj that's awesome! I think I've addressed all your comments so far now 😄

a8ck3n commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+		// get count as a percent value
+		return ( segmentCount / totalCount ) * 100;
+	};
+
```
Should we add some bounds checking on our props here? For example, making sure the segmentCount is less than the totalCount, keeping thickness between 1 & some upper bound, setting a minimum donutWidth?

annacmc replied:
Yes, absolutely! If it gets to thick, for example, it stops rounding out so nicely. It would be excellent to prevent that. And also checking the segmentCount and totalCount could be checked here for sure. The decision surrounding this would be deciding what to show if the segmentCount was, for example, over the totalCount. Depending on how the new billing API data comes through, it may be possible that someone is over their limit and we would want to display the chart just at 100%.

a8ck3n commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+	};
+
+	return (
+		<div className="donut-meter" aria-hidden="true">
```
Instead of hiding the chart from screen readers, maybe we can mark it up a bit more?

- Wrap the DIV in a FIGURE element.
- Add ARIA-LABELLEDBY element.
- Add TITLE & DESCRIPTION elements.
- Make the TITLE & DESCRIPTION values props.
- Set the role attribute to IMG.

```
<figure>
  <div class="figure-content">
    <svg width="100%" height="100%" viewBox="0 0 42 42" class="donut" aria-labelledby="beers-title beers-desc" role="img">
      <title id="beers-title">Beers in My Cellar</title>
      <desc id="beers-desc">Donut chart showing 10 total beers. Two beers are Imperial India Pale Ales, four beers are Belgian Quadrupels, and three are Russian Imperial Stouts. The last remaining beer is unlabeled.</desc>
```

This might eliminate the need for the summary table following the meter as well.

annacmc replied:
I like this idea a lot. Noting that this table is almost identical to the RecordMeterBar component chart which serves the same purpose. So it might be worth dropping into that component and making the same changes 😄

a8ck3n commented on projects/js-packages/components/components/record-meter-donut/index.tsx:
```diff
@@ -0,0 +1,104 @@
+						<td>{ totalCount }</td>
+					</tr>
+				</tbody>
+			</table>
```
I think we should make the screen reader text more descriptive. As it currently stands, the screen reader is only announcing the meter and the values but without any context. If we have two meters side by side, it might not be clear what they represent. I think these should be strings that are passed to us.

As noted above, I think it makes sense to put this inline using the TITLE & DESCRIPTION elements instead of in an associated table though. Maybe the table approach was used in the RecordMeter due to it being an inaccessible CANVAS element?

Either way, it would then be caller's responsibility to generate the correct localized strings and pass them in via props. We could then remove the dependency on `/i18n` entirely.

annacmc replied:
>Maybe the table approach was used in the RecordMeter due to it being an inaccessible CANVAS element?

For some background/context here, the table was initially used in the first RecordMeter due to it being an inaccessible canvas element. Then, when the RecordMeterBar component was built, it had no accessibility fallbacks so we threw the table back in to make it a bit more accessibility friendly. I do agree it could be improved upon in both the RecordMeterBar and the DonutMeter components. You can see the [PR for when we added the table to RecordMeterBar here](https://github.com/Automattic/jetpack/pull/24748)

a8ck3n reviewed (commented):
This is looking awesome, @annacmc. I love it!

I shared a few suggestions around inputs, localization, and a11y.

I'm also wondering if we should break out the segment value into an array so we can support more than one segment. I know you asked about this when we were discussing in Slack. Now that I'm seeing how the code works, I think it might be worth adding this. I'm thinking we could pass in an array of [percent, colour, title, desc] values and generate the segments from that data. What do you think?

PS: I realize you won't be around to work on this for the next six month so maybe this is something we could tackle in a second round of tweaks. Or maybe we take the excellent base you have here and continue running with it.

oskosk reviewed (approved):
LGTM

I suggest going forward and merging the current state of this PR... 

Then coming back to address observations on follow-up PRs. 

Added a tracking issue for the remaining observations

[JS Components: DonutMeter - Follow up on recommendations from #25947](https://github.com/Automattic/jetpack/issues/26049)

oskosk commented:
r252030-wpcom
